bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -220,16 +224,49 @@ pub mod pallet {
 			origin: OriginFor<T>,
 			asset_id: Box<VersionedLocation>,
 			metadata: AssetMetadata,
+			fee_asset: Asset,
 		) -> DispatchResult {
 			ensure!(!Self::export_operating_mode().is_halted(), Error::<T>::Halted);
 
 			let asset_location: Location =
 				(*asset_id).try_into().map_err(|_| Error::<T>::UnsupportedLocationVersion)?;
 			let origin_location = T::RegisterTokenOrigin::ensure_origin(origin, &asset_location)?;
+			let who = T::AccountIdConverter::convert_location(&origin_location)
+				.ok_or(Error::<T>::LocationConversionFailed)?;
+
+			let ether_location = T::EthereumLocation::get();
+			let (fee_asset_location, fee_amount) = match fee_asset {
+				Asset { id: AssetId(ref loc), fun: Fungible(amount) } => (loc, amount),
+				_ => {
+					tracing::debug!(target: LOG_TARGET, ?fee_asset, "error matching fee asset");
+					return Err(Error::<T>::UnsupportedAsset.into())
+				},
+			};
+			ensure!(fee_amount > 0, Error::<T>::TipAmountZero);
+
+			let ether_gained = if *fee_asset_location != ether_location {
+				Self::swap_and_burn(
+					who.clone(),
+					fee_asset_location.clone(),
+					ether_location,
+					fee_amount,
+				)
+				.inspect_err(|&e| {
+					tracing::debug!(target: LOG_TARGET, ?e, "error swapping asset");
+				})?
+			} else {
+				burn_for_teleport::<T::AssetTransactor>(&origin_location, &fee_asset)
+					.map_err(|_| Error::<T>::BurnError)?;
+				fee_amount
+			};
```

Comment:
Since this code is duplicated in `add_tip`, perhaps we should extract it to a method?

Reply:
https://github.com/paritytech/polkadot-sdk/pull/8725/commits/d2f43b9467e623cffe5a99a3321f70c4f23cbc4a